### PR TITLE
[CALCITE-3939] Change UnionEliminatorRule and ProjectRemoveRule to auto pruning SubstitutionRule

### PR DESCRIPTION
**Pull request by hbtoo**
https://issues.apache.org/jira/browse/CALCITE-3939

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectRemoveRule.java [hunk 1]
```diff
@@ -88,7 +88,6 @@ public static boolean isTrivial(Project project) {
         project.getInput().getRowType());
   }
 
-  @Deprecated // to be removed before 1.5
   public static boolean isIdentity(List<? extends RexNode> exps,
       RelDataType childRowType) {
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectRemoveRule.java:92]: I don't understand what you want to do. Could you explain more about it?

Comment (hbtoo) [reply core/src/main/java/org/apache/calcite/rel/rules/ProjectRemoveRule.java]: I am only testing the continuous testing failure with a trivial change. It seems irrelevant to my patch. I've just pushed the real change.

File: core/src/main/java/org/apache/calcite/rel/rules/UnionEliminatorRule.java [hunk 2]
```diff
@@ -44,18 +46,17 @@ public UnionEliminatorRule(Class<? extends Union> clazz,
...
-    if (union.getInputs().size() != 1) {
-      return;
-    }
-    if (!union.all) {
-      return;
-    }
-
-    // REVIEW jvs 14-Mar-2006:  why don't we need to register
-    // the equivalence here like we do in AggregateRemoveRule?
+    return union.all && union.getInputs().size() == 1;
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/rel/rules/UnionEliminatorRule.java:51]: I am curious when can we get a union with only 1 child?

Comment (hbtoo) [reply core/src/main/java/org/apache/calcite/rel/rules/UnionEliminatorRule.java:51]: I guess one possible case is this. Unions are typically created via:
public static LogicalUnion create(List<RelNode> inputs, boolean all);

"List<RelNode> inputs" is typically assembled with a for loop. Whenever the for loop executes with only one loop, we have an input list of one element.

**Pull request closed without merging**